Automattic/wp-calypso
# Gutenberg: make sure that editor initialization is executed only once

Fixes https://github.com/Automattic/wp-calypso/issues/27882 and https://github.com/Automattic/wp-calypso/issues/27993

#### Changes proposed in this Pull Request

The problem with our current editor implementation is that whenever we navigate away from it (by clicking `Close`) and then navigate back, all of the initialization logic will be executed again. That means that API middleware will be applied twice, core blocks registration will be attempted again (producing warnings) etc.

I discussed this with @Automattic/team-calypso in https://github.com/Automattic/wp-calypso/issues/27882 and @jsnajdr [proposed two options](https://github.com/Automattic/wp-calypso/issues/27882#issuecomment-430582147):

1. placing the init in the default export function of `client/gutenberg/editor/index.js`.
2. placing the init in the `post` controller wrapped with `lodash/once`.

After experimenting a bit with both approaches, I decided to go with option 2. It seems more readable and easier to reason about to me, because all of the setup logic and sequencing is in one place. We also don't have to pull in `redux-bridge` for it.

Here is the summary of what has been moved to new initialization function:

- All of the init code from `main.jsx` - This includes: data plugins registration, `core-data` init, API middleware setup, Classic Block init, Core Blocks registration, and TinyMCE plugins init.
- The API middleware has been refactored and the utility component has been removed. Previously, this component was ensuring that API middleware was applied before the editor starts firing API requests. However, it was not accounting for the above issue, so same middleware was applied multiple times. I suspect this didn't surface earlier because we are breaking the middleware chain - if that was not the case our mapping would be broken after navigating away and back. 
- I moved the editor extension loading (presets) to init function too, since it seems like a better place to contain that functionality. Previously it was placed in `edit-post` folder, and we'd like to avoid adding Calypso's custom modifications there as much as possible. Maintaining it is hard enough with frequent core changes, and adding our custom code there only makes it harder. We are working to move away from it as soon as possible, but it's not an immediate priority.

#### Testing instructions

* `localStorage.setItem( 'debug', 'calypso:gutenberg' );`
* Navigate to http://calypso.localhost:3000/gutenberg/post/
* Notice the editor initialization debug messages in console.
* Click `Close`.
* Click `Back` in your browser (at this point there is a similar visual glitch that we observed in Calypsoify redirects, not related to this PR and present in wpcalypso too).
* Verify that editor initialization has not been executed twice by observing the console log. Currently in `wpcalypso` you'd see repeated core blocks registration warnings.
* Smoke test the editor and verify that this is not causing any regressions.
